internal/transport/http_util_test.go

```diff
@@ -185,3 +188,56 @@ func (s) TestDecodeMetadataHeader(t *testing.T) {
 		}
 	}
 }
+
+func (s) TestDecodeHeaderH2ErrCode(t *testing.T) {
+	for _, test := range []struct {
+		name string
+		// input
+		metaHeaderFrame *http2.MetaHeadersFrame
+		serverSide      bool
+		// output
+		h2code http2.ErrCode
+	}{
+		{
+			name: "valid header",
+			metaHeaderFrame: &http2.MetaHeadersFrame{Fields: []hpack.HeaderField{
+				{Name: "content-type", Value: "application/grpc"},
+			}},
+			h2code: http2.ErrCodeNo,
+		},
+		{
+			name: "invalid grpc status header field",
+			metaHeaderFrame: &http2.MetaHeadersFrame{Fields: []hpack.HeaderField{
+				{Name: "content-type", Value: "application/grpc"},
+				{Name: "grpc-status", Value: "xxxx"},
+			}},
+			h2code: http2.ErrCodeProtocol,
+		},
+		{
+			name: "invalid http content type",
+			metaHeaderFrame: &http2.MetaHeadersFrame{Fields: []hpack.HeaderField{
+				{Name: "content-type", Value: "application/json"},
+			}},
+			h2code: http2.ErrCodeProtocol,
+		},
+		{
+			name: "http fallback and invalid http status",
+			metaHeaderFrame: &http2.MetaHeadersFrame{Fields: []hpack.HeaderField{
+				// No content type provided then fallback into handling http error.
+				{Name: ":status", Value: "xxxx"},
+			}},
+			h2code: http2.ErrCodeProtocol,
+		},
+		{
+			name:            "http2 frame size exceeds",
+			metaHeaderFrame: &http2.MetaHeadersFrame{Fields: nil, Truncated: true},
+			h2code:          http2.ErrCodeFrameSize,
+		},
+	} {
+		state := &decodeState{serverSide: test.serverSide}
+		h2code, _ := state.decodeHeader(test.metaHeaderFrame)
+		if h2code != test.h2code {
+			t.Fatalf("decodeState.decodeHeader, %q, want %v, got %v", test.name, test.h2code, h2code)
```

Comment:
You could also switch to using `t.Run`, thereby you wont have to specify `test.name` as an argument here. `t.Run` automatically does that.

Also, it would be helpful to add the `metaHeaderFrame` to the error output.
```
t.Fatalf("decodeState.decodeHeader(%v) = %v, want %v", test.metaHeaderFrame, h2code, test.wantCode)
```